uqcsbot/scripts/crates.py

```diff
@@ -0,0 +1,561 @@
+    :param sort: The order to sort by. One of "crates" or "alpha"
+    :param page: The page number to get the categories from
+    :return: A tuple containing a list of category names (or None on error) and the total number of categories
+    """
+    # Get the categories
+    url = BASE_URL + '/categories'
+    response = requests.get(url, {'sort': sort, 'page': page})  # type: ignore
+
+    if response.status_code != requests.codes.ok:
+        bot.post_message(channel, 'There was a problem getting the list of categories')
+        return None, 0
+
+    # Convert the json response
+    response_data = json.loads(response.content)
+
+    raw_categories = response_data.get('categories')
+    total = response_data.get('meta', {}).get('total', 0)
+
+    # Get the category names
+    categories = [cat.get('name') if 'name' in cat else cat.get('id', '') for cat in raw_categories]
+
+    return categories, total
+
+
+def display_all_categories(channel: Channel, args: CategorySearch):
+    """Displays just the names of all the categories in one big list"""
+    cats, tot = get_category_page(channel, args.sort, 1)
+    if cats is None:
+        return  # Error occurred
+
+    # Get all of the categories by incrementing page number
+    page = 2
+    while len(cats) < tot:
+        next_cats, _ = get_category_page(channel, args.sort, page)
+        if next_cats is None or not next_cats:
+            break
+
+        cats.extend(next_cats)
+        page += 1
+
+    # Begin formatting the message
+    category_string = '\n'.join(cats)
+    blocks = [
+        {
+            'type': 'section',
+            'text': {
+                'type': 'mrkdwn',
+                'text': f'*Displaying {tot} categories:*'
+            }
+        },
+        {
+            'type': 'section',
+            'text': {
+                'type': 'mrkdwn',
+                'text': f'```{category_string}```'
+            }
+        },
+    ]
+
+    bot.post_message(channel, '', blocks=blocks)
+
+
+def display_specific_category(channel: Channel, args: CategorySearch):
+    """Displays a single category in more detail"""
+    # Get the categories
+    url = BASE_URL + f'/categories/{args.name}'
+    response = requests.get(url)
+
+    if response.status_code != requests.codes.ok:
+        bot.post_message(channel, f'There was a problem getting the category "{args.name}"')
+        return
+
+    # Convert the json response
+    response_data = json.loads(response.content)
+    if 'errors' in response_data:
+        bot.post_message(channel, f'The category "{args.name}" does not exist')
+        return
+
+    raw_category = response_data.get('category')
+
+    name = raw_category.get('name')
+    name = raw_category.get('id') if name is None else name
+    desc = raw_category.get('description', 'No description provided')
+
+    category = CategoryResult(name, desc, raw_category['crates_cnt'])
+
+    # Format the message
+    blocks = [
+        {
```

Comment:
These dictionaries look like the kind of thing that could be produced algorithmically. Just a thought...

Reply:
I have added some classes and functions to do this in a rather simplistic way however I feel as though it doesn't really belong in a script file.
I think the bot might benefit from having something dedicated to constructing responses using the new blocks layout as it does seem to fit well with modular message construction.

Reply:
You're right, we should push this upwards. Want to make an issue and link back to this?